### ISSUE TITLE
Add NetworkView Control

### ISSUE THREAD

**Issue by insinfo** [Microsoft/xaml-standard#163]
I believe that everyone in XAML lovers, we want a robust and complete standard and that facilitates the life of the developer so I propose to include these controls as a standard of XAML specification. These are widely used in the real world today. In several areas we find these types of control such as audio/video playes, visual programming, animation software, single/multitracks audio editors, video editors, slide editors, photo gallery, websites, social networks among many others.

A  control for visualizing and editing networks, graphs and flow-charts

**# NetworkView Control**
```xaml
<NetworkView>
    <Node Name="node1" Id="node1" Class="" Title="" Icon="" Info="">
        <Connections>
            <InPort PortId="125" ConnectionPortId="255" Title=""/>
            <OutPort PortId="2500" ConnectionPortId="150" Title=""/>
        </Connections>
    </Node>
    <Node Name="node2" Id="node2" Class="" Title="" Icon="" Info="">
        <Connections>
            <InPort PortId="125" ConnectionPortId="255" Title=""/>
            <OutPort PortId="2500" ConnectionPortId="150" Title=""/>
        </Connections>
    </Node>
</NetworkView>
```
https://www.codeproject.com/Articles/182683/NetworkView-A-WPF-custom-control-for-visualizing-a

**Comment by insinfo** [comment]
I updated with other demos on Android, IOS and UWP

**Comment by insinfo** [comment]
I am still hopeful that a great XAML standard will be created here, complete, robust and that will make life easier for the developer.
You can not only think of generic controls, generic controls are super important they are the basis but you can not stop there. If you want to create a real-world application easily and quickly, you can not waste time reinventing the wheel all the time, having to create custom controls for trivial things already well-known and routinely used by millions of applications. I believe that a standard one-language graphical interface creation has to be well complete, robust, extendable most above all to make the developer's life to the fullest.

**Comment by dotMorten** [comment]
`<rant>` @insinfo you need to stop this. You posted the same message in multiple issues and off topic AGAIN. This isn't a chat forum. We log an issue and keep the discussion in each issue on point and on topic. Every morning I have a gazillion email notifications from this repo because you verbal vomited all over it. You're only hurting your own issues acting like this. Please learn some github etikette and try and better understand what the spec is about rather than steer it away from what it is. `</rant>`

**Comment by birbilis** [comment]
@dotMorten indeed, probably @insinfo doesn't use e-mail notifications. The reply might originally seem to them as applicable to each of those posts, but receiving the exact same test via multiple e-mail notifications is like spam. Probably GitHub could catch this activity and warn one against it in the first place

**Comment by Mike-EEE** [comment]
YOU CANNOT HOPE TO STOP THE POWER OF @insinfo, YOU CAN ONLY PRAY TO CONTAIN IT!!!

![](https://media.giphy.com/media/yUI3a7RwLhOFy/giphy.gif)

(Sorry, I keep saying this to myself whenever I get a new notification that we are so affectionately discussing.)

**Comment by birbilis** [comment]
related issue is Adorners, people where using them in WPF to make connectors appear between visual controls (and wiring) without needing a special container (that is suggested here instead) for those controls
https://github.com/Microsoft/xaml-standard/issues/101

a special control of course can help with persisting such a connectivity graph and related properties of that graph, but it could be a control that is given another container as a child or even better allow giving it by name too via a property without the need to always nest it and it injects this wiring (and persistence of wiring) functionality using adorner-based UI

**Comment by birbilis** [comment]
speaking of Adorners, the article you mentioned above for the WPF NetworkView seems to also use them:

>  ZoomAndPanControl is from my earlier article and I won't be going into detail on it here. Suffice to say that it gives the ability to zoom and pan the NetworkView.
>
> AdornerDecorator creates a new adorner layer underneath the ZoomAndPanControl in the visual-tree. Adorners are used to display the delete connection and delete node mouse-hover buttons and also the feedback icons. Without the AdornerDecorator then the adorners would be displayed within the default adorner layer that is above the ZoomAndPanControl in the visual-tree. If this adorner layer were used instead of the explicitly defined one it would mean that the adorners would not be subject to the scale and offset transformations that are applied by the ZoomAndPanControl (via its render transform) and therefore would not appear at the same zoom level as the rest of ZoomAndPanControl's content.
>
> dragZoomCanvas (after the NetworkView) is used to render the zoom rectangle that allows the user to drag out a rectangle to zoom to. This technique was described in the ZoomAndPanControl article.

**Comment by crutkas** [comment]
Everyone please let's stay on topic.  New issue please keep it there.  Scenarios and use cases on how YOU do stuff is critical.

**Comment by KySpace** [comment]
Before that I would suggest to add the FreeBox control, that the boxes can float and adjust their positions, and then have the Network connecting them together. Also please support properties like notations that are on this connections, and connection style etc.

**Comment by insinfo** [comment]
@KySpace
Thanks for looking at this control, I believe that this control is very important for a robust and complete XAML standard, and for modern next-generation software that involves visual programming, interactivity with neural networks and other purposes, you could give an example in XAML of your suggestion, so I can update the initial post with your suggestion.

**Comment by insinfo** [comment]
@KySpace 
And do not forget to vote, a smile +1

**Comment by birbilis** [comment]
There are some nice libraries for Graphs (say GraphX for .NET and MSAGL from MS Research etc.) out there that could be examined. A common feature they have is that layout algorithms are available for choosing and or plugging in.

Here's a collection of links on Visualization (with a Graphs subfolder) that might be helpful:
https://www.dropbox.com/sh/rbkrbi4vw2j7n4d/AACxwpQ0unKwkwujbuRLWW-la?dl=0

Those links where collected while doing research for the ClipFlair project, however never ended up adding a Graph component to it, due to the extra time needed to port GraphX (which was my 1st choice) to Silverlight back then (the usecase was to allow an educator/student to add mind maps and flowchart style guides etc. to their FLL activities)

**Comment by insinfo** [comment]
@birbilis it seems that most of the links you posted are no longer available

**Comment by insinfo** [comment]
@birbilis 
> make sure that the base player has extensibility so that MMPPF can be build upon it and be portable to more platforms

I fully agree with you

**Comment by birbilis** [comment]
@insinfo unfortunately the Internet is becoming more and more a swamp of broken links. Some that come to mind are silverlight.net forums broken links, or Codeplex going into archive mode, MS retiring older Visual Studio documentation into downloadable files breaking its own articles that point to them etc.

Luckily there is http://archive.org that archives some of that stuff and one can put a URL there to see snapshots from older dates if they're looking for some content. Of course they can't store everything, wish they keep on their effort though
